Python/chapter02/2.8 - Loop Detection/miguel_soln_2.8.py

```diff
@@ -164,6 +164,31 @@ def __eq__(self, other: object):
         return a is None and b is None
 
 
+def loop_detection_const_space(ll: LinkedList) -> Optional[Node]:
+    """
+    This function will determine if there is a
+    cycle in the input linked list.
+    A linked list is 'circular' when a node's
+    next pointer points to an earlier node, so
+    as to make a loop in the linked list.
+    Runtime:  O(n^2)
+    Space Complexity:  O(1)
+    :param ll: an input linked list
+    :return: the corrupt node or None
+    """
+    if ll.head and ll.head.next is ll.head:
+        return ll.head
+    n1 = ll.head
+    while n1 is not None:
+        n2 = ll.head
+        while n2 is not n1:
```

Comment:
what happens if you add a test case with a very long chain of elements and a very long cycle? (say, 10 items each)

also, can you modify the test generation such that you only manipulate the `.next` pointers instead of using the append functions? there's a bug here that's not being caught by the tests :S

Reply:
long cycle test case added.

I do not see the bug that is not being caught by the tests.  Could you give me another hint?

Reply:
after more careful inspection, it was just me misreading the code u__u.

i was misinterpreting what `n1` and `n2` meant. can you add a few comments about how this all works and maybe also rename those variables to avoid future readers from also being confused?

Reply:
Yes, good idea!

Reply:
updated